javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java

```diff
@@ -0,0 +1,139 @@
+import com.sun.source.tree.TryTree;
+import com.sun.source.tree.UnaryTree;
+import com.sun.source.tree.VariableTree;
+import com.sun.tools.javac.tree.JCTree;
+import com.sun.tools.javac.tree.JCTree.JCAssignOp;
+import com.sun.tools.javac.tree.JCTree.JCBinary;
+import com.sun.tools.javac.tree.JCTree.JCClassDecl;
+import com.sun.tools.javac.tree.JCTree.JCExpression;
+import com.sun.tools.javac.tree.JCTree.JCFieldAccess;
+import com.sun.tools.javac.tree.JCTree.JCIdent;
+import com.sun.tools.javac.tree.JCTree.JCLambda;
+import com.sun.tools.javac.tree.JCTree.JCMemberReference;
+import com.sun.tools.javac.tree.JCTree.JCMethodDecl;
+import com.sun.tools.javac.tree.JCTree.JCMethodInvocation;
+import com.sun.tools.javac.tree.JCTree.JCNewClass;
+import com.sun.tools.javac.tree.JCTree.JCTry;
+import com.sun.tools.javac.tree.JCTree.JCUnary;
+import com.sun.tools.javac.tree.JCTree.JCVariableDecl;
+import com.sun.tools.javac.tree.TreeCopier;
+import com.sun.tools.javac.tree.TreeMaker;
+
+/** A Utility class for coping AST {@link JCTree} fully, including type and symbol information. */
+class FullyTreeCopier extends TreeCopier<Void> {
+
+    private TreeMaker M;
+
+    FullyTreeCopier(TreeMaker M) {
+        super(M);
+        this.M = M;
+    }
+
+    public <T extends JCTree> T copy(T tree, Void p) {
+        T t = super.copy(tree, p);
+        if (t == null) {
+            return null;
+        }
+
+        t.type = tree.type;
+        return t;
+    }
+
+    public JCTree visitCompoundAssignment(CompoundAssignmentTree node, Void p) {
+        JCAssignOp t = (JCAssignOp) super.visitCompoundAssignment(node, p);
+        t.operator = ((JCAssignOp) node).operator;
+        return t;
+    }
+
+    public JCTree visitBinary(BinaryTree node, Void p) {
+        JCBinary t = (JCBinary) super.visitBinary(node, p);
+        t.operator = ((JCBinary) node).operator;
+        return t;
+    }
+
+    public JCTree visitClass(ClassTree node, Void p) {
+        JCClassDecl t = (JCClassDecl) super.visitClass(node, p);
+        t.sym = ((JCClassDecl) node).sym;
+        return t;
+    }
+
+    public JCTree visitIdentifier(IdentifierTree node, Void p) {
+        JCIdent t = (JCIdent) super.visitIdentifier(node, p);
+        t.sym = ((JCIdent) node).sym;
+        return t;
+    }
+
+    public JCTree visitMethod(MethodTree node, Void p) {
+        JCMethodDecl t = (JCMethodDecl) super.visitMethod(node, p);
+        t.sym = ((JCMethodDecl) node).sym;
+        return t;
+    }
+
+    public JCTree visitMethodInvocation(MethodInvocationTree node, Void p) {
+        JCMethodInvocation t = (JCMethodInvocation) super.visitMethodInvocation(node, p);
+        t.varargsElement = ((JCMethodInvocation) node).varargsElement;
+        return t;
+    }
+
+    public JCTree visitNewClass(NewClassTree node, Void p) {
+        JCNewClass t = (JCNewClass) node;
+        JCExpression encl = (JCExpression) super.visitNewClass(node, p);
+        t.constructor = ((JCNewClass) node).constructor;
```

Comment:
Isn't this a no-op? `t` is an alias for `node` and here you are casting `node` again.
Am I missing what all these assignments do?